### PR TITLE
add: unit test

### PR DESCRIPTION
**Pull request by kavyakunder**
test for card
tets for card app

### CODE REVIEW

File: src/App.test.tsx [hunk 1]
```diff
@@ -1,9 +1,91 @@
...
+import { CardProps } from "./component/Card";
+
+jest.mock("./component/Card", () => ({
+  Card: ({ card, handleCardClick }: CardProps) => (
+    <div onClick={() => handleCardClick(card)}>Card</div>
+  ),
+}));
+
+describe("App", () => {
+  it("renders title,subtitle and button", async () => {
```
Comment (shah-harshit) [inline src/App.test.tsx]:
```suggestion
  it("should render title, subtitle and buttons", async () => {
```

File: src/App.test.tsx [hunk 2]
```diff
@@ -1,9 +1,91 @@
...
+    render(<App />);
+    const title = screen.getByTestId("title");
+    const subtitle = screen.getByTestId("sub-title");
+    expect(title).toBeInTheDocument();
```
Comment (shah-harshit) [inline src/App.test.tsx]: In this case, we can move all expects together at the bottom.

File: src/App.test.tsx [hunk 3]
```diff
@@ -1,9 +1,91 @@
...
+    fireEvent.click(allCards[0]);
+    expect(allCards[0].textContent).not.toBe("");
+
+    fireEvent.click(allCards[1]);
+    expect(allCards[1].textContent).not.toBe("");
+
+    const moves = screen.getByTestId("moves");
+    expect(moves).toHaveTextContent("Moves: 1");
+
+    await act(async () => {
```
Comment (shah-harshit) [inline src/App.test.tsx]: Since we are not using await for `jest.runAllTimers();`. We don't need async here.

File: src/App.test.tsx [hunk 4]
```diff
@@ -1,9 +1,91 @@
...
+    const btnEasy = screen.getByTestId("btn-easy");
+    const btnMedium = screen.getByTestId("btn-medium");
+    const btnHard = screen.getByTestId("btn-hard");
+
+    expect(btnEasy).toBeInTheDocument();
+    expect(btnMedium).toBeInTheDocument();
+    expect(btnHard).toBeInTheDocument();
+  });
+
+  it("displays 8 cards when clicked on easy", async () => {
```
Comment (shah-harshit) [inline src/App.test.tsx]:
```suggestion
  it("Should display 8 cards when clicked on easy", async () => {
```

Please make this changes for all the test titles.

File: src/component/Card.test.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    expect(card.innerHTML).toContain(flippedCard.design);
+  });
+
+  it("shows the back design when the card is not flipped", () => {
+    render(<Card {...mockProps} />);
+
+    const card = screen.getByTestId("card");
+    expect(card.innerHTML).not.toContain(mockCard.design);
+  });
+});
```
Comment (shah-harshit) [inline src/component/Card.test.tsx]: We don't need this tests.

File: src/App.test.tsx [hunk 6]
```diff
@@ -1,9 +1,91 @@
...
+    fireEvent.click(allCards[1]);
+
+    const moves = screen.getByTestId("moves");
+    expect(moves).toHaveTextContent("Moves: 1");
+
+    act(() => {
+      jest.runAllTimers();
+    });
+
+    expect(allCards[0]).toHaveTextContent("Card");
```
Comment (shah-harshit) [inline src/App.test.tsx]: I don't think we need this.

File: src/App.test.tsx [hunk 7]
```diff
@@ -1,9 +1,91 @@
...
+
+    fireEvent.click(allCards[0]);
+    fireEvent.click(allCards[1]);
+
+    const moves = screen.getByTestId("moves");
+    expect(moves).toHaveTextContent("Moves: 1");
+
+    act(() => {
+      jest.runAllTimers();
+    });
```
Comment (shah-harshit) [inline src/App.test.tsx:72]: Can you try is the test is working without this?

File: src/App.test.tsx [hunk 8]
```diff
@@ -1,9 +1,91 @@
...
+    render(<App />);
+    const btnHard = screen.getByTestId("btn-hard");
+    fireEvent.click(btnHard);
+
+    const btnRestart = screen.getByTestId("btn-restart");
+    const moves = screen.getByTestId("moves");
+    expect(btnRestart).toBeInTheDocument();
+    expect(moves).toBeInTheDocument();
+
+    fireEvent.click(btnRestart);
```
Comment (shah-harshit) [inline src/App.test.tsx:90]: Need an assertion to verify if the click on restart button was successful.

File: src/App.test.tsx [hunk 9]
```diff
@@ -1,9 +1,91 @@
...
+import { App } from "./App";
+import { CardProps } from "./component/Card";
+
+jest.mock("./component/Card", () => ({
+  Card: ({ card, handleCardClick }: CardProps) => (
+    <div onClick={() => handleCardClick(card)}>Card</div>
+  ),
+}));
+
+describe("App", () => {
```
Comment (shah-harshit) [inline src/App.test.tsx]:
```suggestion
describe("App.tsx", () => {
```

File: src/component/Card.test.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import { Card } from "./Card";
+
+import { fireEvent, render, screen } from "@testing-library/react";
+import { CardProps } from "./Card";
+
+describe("Card", () => {
```
Comment (shah-harshit) [inline src/component/Card.test.tsx]:
```suggestion
describe("Card.tsx", () => {
```

File: src/App.test.tsx [hunk 11]
```diff
@@ -1,9 +1,91 @@
...
+  });
+
+  it("Should restart the game", () => {
+    render(<App />);
+    const btnHard = screen.getByTestId("btn-hard");
+    fireEvent.click(btnHard);
+
+    const btnRestart = screen.getByTestId("btn-restart");
+    const moves = screen.getByTestId("moves");
+    expect(btnRestart).toBeInTheDocument();
```
Comment (shah-harshit) [inline src/App.test.tsx:87]: Need to add a line between const and expect here.

File: src/App.test.tsx [hunk 12]
```diff
@@ -1,9 +1,91 @@
...
+
+    act(() => {
+      jest.runAllTimers();
+    });
+
+    expect(allCards[0]).toHaveTextContent("Card");
+  });
+
+  it("Should restart the game", () => {
+    render(<App />);
```
Comment (shah-harshit) [inline src/App.test.tsx:79]: Need to add a line between const and render here.

File: src/App.test.tsx (hunk 7 above)
Comment (kavyakunder) [reply src/App.test.tsx:72]: test is working, but coverage will not  be 100%

Comment (shah-harshit) [reply src/App.test.tsx:72]: Are you sure? I don't think jest.runAllTimers has anything to do with coverage.

**Review by shah-harshit — APPROVED** [review]

**Pull request merged**